src/plugins/grpc/server.js

```diff
@@ -0,0 +1,137 @@
+      const { type } = this
+      const isStream = ['client_stream', 'server_stream'].includes(type)
+      const childOf = tracer.extract(FORMAT_HTTP_HEADERS, metadata.getMap())
+      const span = tracer.startSpan('grpc.server', {
+        childOf,
+        tags: {
+          [Tags.SPAN_KIND]: Tags.SPAN_KIND_RPC_SERVER,
+          'grpc.method_name': handler,
+          'grpc.method_type': type,
+          'resource.name': handler.split('/').slice(-1),
+          'service.name': config.service || `${tracer._service}-grpc`
+        }
+      })
+
+      tracer.scopeManager().activate(span)
+
+      if (request) {
+        if (metadata) {
+          for (const [key, value] of Object.entries(metadata.getMap())) {
+            span.setTag(`grpc.metadata.${key}`, value)
+          }
+        }
+
+        if (config.arguments) {
+          const values = config.arguments(request)
+
+          for (const key in values) {
+            span.setTag(`grpc.arguments.${key}`, values[key])
+          }
+        } else {
+          span.setTag('grpc.arguments', request)
+        }
+      }
+
+      if (isStream) {
+        emitter.on('error', err => {
+          handleError(config, span, err)
+
+          span.finish()
+        })
+
+        // Finish the span if the response only if it was successful.
+        // Otherwise it'll be finished in the `error` listener.
+        emitter.on('finish', function () {
+          if (emitter.status.code === 0) {
+            span.finish()
+          }
+        })
+
+        // Call the original stream request, without modification.
+        return func.apply(this, arguments)
+      }
+
+      const [, callback] = arguments
+
+      // Call the unary request with a wrapped callback.
+      return func.call(this, emitter, (err, value, trailer, flags) => {
+        if (err) {
+          handleError(config, span, err, trailer)
+        }
+
+        span.finish()
+
+        if (callback) {
+          callback.call(this, value, trailer, flags)
+        }
+      })
+    }
+  }
+}
+
+function handleError (config, span, err, metadata) {
+  span.addTags({
+    'error.code': err.code,
+    'error.msg': err.message,
+    'error.stack': err.stack
+  })
+
+  if (metadata) {
+    if (config.arguments) {
+      const values = config.arguments(metadata.getMap())
+
+      for (const key in values) {
+        span.setTag(`error.metadata.${key}`, values[key])
+      }
+    } else {
+      span.setTag(`error.metadata`, metadata.getMap())
+    }
+  }
+}
+
+function patch (grpc, tracer, config) {
+  if (config.server === false) return
+
+  config = config.server || config
+
+  const self = this
+
+  this.wrap(grpc.Server.prototype, 'start', start => {
+    return function () {
+      start.call(this)
+
+      for (const handler in this.handlers) {
+        self.wrap(this.handlers[handler], 'func', createWrapHandler(grpc, tracer, config, handler))
+
+        patchedHandlers.push([this.handlers[handler], 'func'])
```

Comment:
Using an object would make this clearer.

Reply:
Also, why is the name needed if it's always `func`?

Reply:
> Using an object would make this clearer.

I gave it a try but I find it a bit more unclear:

```js
patchedHandlers[handler] = this.handlers[handler]
```

Also, it forces me to iterate over the object's values in the `unpatch` method and completely discard the object's keys. (which is why I went for an array 😇 )

Do you have a suggestion here?

> Also, why is the name needed if it's always func?

I was initially patching multiple functions and needed a generic solution but have since managed to find the `func` injection point. I'll hard-code in the unpatch method. 👍

Reply:
> Also, it forces me to iterate over the object's values in the unpatch method and completely discard the object's keys.

I meant something like `patchedHandlers.push({ handler, name })`. Then when you iterate it's easier to understand as you are accessing a property with proper semantics instead of an index.

This may now be a non-issue though if you are going to remove the name since it's no longer necessary.